File: Low Level Design (LLD)/Problems/06. Splitwise/Java/model/Expense.java
package model;

import model.enums.ExpenseType;

import java.util.*;

public abstract class Expense {
    protected String expenseId;
    protected String description;
    protected double amount;
    protected User paidBy;
    protected List<User> participants;
    protected ExpenseType type;
    protected Date createdAt;
    protected Map<User, Double> userShare;

    public Expense(String expenseId, String description, double amount, User paidBy, List<User> participants, ExpenseType type) {
        this.expenseId = expenseId;
        this.description = description;
        this.amount = amount;
        this.paidBy = paidBy;
        this.participants = participants;
        this.type = type;
        this.createdAt = new Date();
        this.userShare = new HashMap<>();
    }

    public abstract void calculateSplit();

    public String getExpenseId() { return expenseId; }
    public String getDescription() { return description; }
    public double getAmount() { return amount; }
    public User getPaidBy() { return paidBy; }
    public List<User> getParticipants() { return participants; }
    public ExpenseType getExpenseType() { return type; }
    public Date getCreatedAt() { return createdAt; }
    public Map<User, Double> getUserShare() { return userShare; }
}

File: Low Level Design (LLD)/Problems/06. Splitwise/Cpp/model/src/Expense.cpp
#include "../include/Expense.h"

Expense::Expense(const std::string& expenseId,
                 const std::string& description,
                 double amount,
                 std::shared_ptr<User> paidBy,
                 const std::vector<std::shared_ptr<User>>& participants,
                 ExpenseType type)
    : expenseId(expenseId),
      description(description),
      amount(amount),
      paidBy(paidBy),
      participants(participants),
      type(type),
      createdAt(std::chrono::system_clock::now()) {}

const std::string& Expense::getExpenseId() const {
    return expenseId;
}

const std::string& Expense::getDescription() const {
    return description;
}

double Expense::getAmount() const {
    return amount;
}

std::shared_ptr<User> Expense::getPaidBy() const {
    return paidBy;
}

const std::vector<std::shared_ptr<User>>& Expense::getParticipants() const {
    return participants;
}

ExpenseType Expense::getExpenseType() const {
    return type;
}

std::chrono::system_clock::time_point Expense::getCreatedAt() const {
    return createdAt;
}

const std::unordered_map<std::shared_ptr<User>, double>& Expense::getUserShare() const {
    return userShare;
}

std::unordered_map<std::shared_ptr<User>, double>& Expense::getUserShareMutable() {
    return userShare;
}
